sys/kern/sysv_shm.c

```diff
@@ -480,14 +496,33 @@ kern_shmat_locked(struct thread *td, int shmid,
 		else
 			return (EINVAL);
 #if __has_feature(capabilities)
+		/*
+		 * XXX: in principle we should be able to use a reservation
+		 * extending before and after the entry to allow arbitrary
+		 * addresses (subject to available space...).
+		 */
 		if (CHERI_REPRESENTABLE_BASE(attach_va, size) != attach_va)
 			return (EINVAL);
-		if (cheri_gettag(shmaddr))
+		if (cheri_gettag(shmaddr)) {
+			/*
+			 * Fixed mapping through a capability only makes
+			 * sense if we're knowingly remapping.
+			 */
+			if ((shmflg & SHM_REMAP) == 0)
+				return (EINVAL);
+			/* XXX: require that a reservation exists. */
+			/* Handle any rounding above */
 			shmaddr = cheri_setaddress(shmaddr, attach_va);
-		else
-			return (EINVAL);	/* XXX support this? */
-		if (!__CAP_CHECK(shmaddr, size))
-		    return (EINVAL);
+			if (!__CAP_CHECK(shmaddr, size))
+				return (EINVAL);
+		} else {
+			/* As with mmap, untagged implies exclusive. */
+			if ((shmflg & SHM_REMAP) != 0)
+				return (EINVAL);
+			shmaddr = cheri_setaddress(td->td_cheri_mmap_cap,
+			    attach_va);
+
```

Comment:
I think the current implementation in the purecap kernel reservations is not taking into account all cases. I think we have two options to fully integrate reservations in this.
1. Use `vm_map_fixed()` for the case where the reservation must already exist. `vm_map_find()` will create a reservation internally for the case in which we are given an hint.
2. Keep current use of only `vm_map_find()` but will need to teach it to distinguish the two cases by forwarding a `MAP_RESERVATION_CREATE` flag in the case we expect that no reservation exists and one must be created.
I tend towards the first one, as I think it is nice to keep the interface separation so that `vm_map_fixed` always expect a reservation to exist, while `vm_map_find` always expects that it does not.

Reply:
I think that makes sense, I do worry that it will increase the size of the diff to upstream more than adding the flag.